ipset.go

```diff
@@ -268,59 +137,224 @@ func (s *IPSet) Ranges() []IPRange {
+type IPSet struct {
+	rr []IPRange // Sorted by IPRange.less
+}
+
+// Clone returns a copy of s that shares no memory with s.
+func (s *IPSet) Clone() *IPSet {
+	return &IPSet{
+		rr: append([]IPRange{}, s.rr...),
+	}
+}
+
+// Ranges returns the minimum and sorted set of IP
+// ranges that covers s.
+func (s *IPSet) Ranges() []IPRange {
+	return append([]IPRange{}, s.rr...)
 }
 
 // Prefixes returns the minimum and sorted set of IP prefixes
 // that covers s.
-// returning a new slice of prefixes that covers all of the given 'add'
-// prefixes with all the 'remove' prefixes removed.
 func (s *IPSet) Prefixes() []IPPrefix {
-	var out []IPPrefix
-	for _, r := range s.Ranges() {
+	out := make([]IPPrefix, 0, len(s.rr))
+	for _, r := range s.rr {
 		out = append(out, r.Prefixes()...)
 	}
 	return out
 }
 
-// ContainsFunc returns a func that reports whether an IP is in s.
-// The returned func operates on a copy of s, so s may be mutated
-// later.
-func (s *IPSet) ContainsFunc() (contains func(IP) bool) {
-	rv := s.Ranges()
-	// TODO(bradfitz): build a faster data structure with
-	// with s.Prefixes()?
-	return func(ip IP) bool {
-		i := sort.Search(len(rv), func(i int) bool {
-			return ip.Less(rv[i].From)
-		})
-		if i == 0 {
+// Equal returns whether s and o represent the same set of IP
+// addresses.
+func (s *IPSet) Equal(o *IPSet) bool {
+	if len(s.rr) != len(o.rr) {
+		return false
+	}
+	for i := range s.rr {
+		if s.rr[i] != o.rr[i] {
 			return false
 		}
-		i--
-		return rv[i].contains(ip)
 	}
+	return true
 }
 
-// Equal reports whether s contains exactly the same IPs as t.
-func (s *IPSet) Equal(t *IPSet) bool {
-	sr := s.Ranges()
-	tr := t.Ranges()
-	if len(sr) != len(tr) {
-		return false
+// Contains returns whether ip is in s.
+func (s *IPSet) Contains(ip IP) bool {
+	for _, r := range s.rr {
+		if r.Contains(ip) {
+			return true
+		}
 	}
-	for i := range sr {
-		if sr[i] != tr[i] {
-			return false
+	return false
+}
+
+// ContainsRange returns whether all IPs in o are in s.
+func (s *IPSet) ContainsRange(o IPRange) bool {
+	for _, r := range s.rr {
+		if o.coveredBy(r) {
+			return true
 		}
 	}
-	return true
+	return false
+}
+
+// ContainsPrefix returns whether all IPs in p are in s.
+func (s *IPSet) ContainsPrefix(p IPPrefix) bool {
+	return s.ContainsRange(p.Range())
+}
+
+// Overlaps returns whether any IP in o is also in s.
+func (s *IPSet) Overlaps(o *IPSet) bool {
+	// TODO: sorted ranges lets us do this in O(n+m)
+	for _, r := range s.rr {
+		for _, or := range o.rr {
+			if r.Overlaps(or) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// OverlapsRange returns whether any IP in o is also in s.
+func (s *IPSet) OverlapsRange(o IPRange) bool {
+	// TODO: sorted ranges lets us do this more efficiently.
+	for _, r := range s.rr {
+		if r.Overlaps(o) {
+			return true
+		}
+	}
+	return false
+}
+
+// OverlapsPrefix returns whether any IP in p is also in s.
+func (s *IPSet) OverlapsPrefix(p IPPrefix) bool {
```

Comment:
Is it worth having all all these API variants, when it's so easy to get a range out of a prefix? (Maybe it is, but I want to make sure we consider the question.)

Reply:
I didn't consider it too hard, honestly. I figured it's cheap for us to provide accessors for all the top-level types in netaddr, rather than require callers to add visual noise.